### ISSUE TITLE
Segmentation fault on MPI_Finalize

### ISSUE THREAD

**Issue by thetwom** [open-mpi/ompi#6757]
### What version of Open MPI are you using? 
4.0.1

### Describe how Open MPI was installed
from source tarball

### Please describe the system on which you are running
* Operating system/version: SUSE Linux Enterprise Desktop 12 SP2
* Computer hardware: Intel(R) Xeon(R) CPU E3-1276 v3 @ 3.60GHz
-----------------------------

## Details of the problem
I get a segmentation fault on MPI_Finalize()  when using openmpi 4.0.1. This did no occur with openmpi 3.1.0. Tried compiling gcc9 and clang8 without success

I can reproduce the problem with the following simple program:
```
#include "mpi.h"

using namespace std;

int main(int argc, char* argv[])
{
    MPI_Init(&argc, &argv);

    cout << "Hello world" << endl;
    MPI_Finalize();
}
```
compiled with:
```
g++ -O0 -g -Ipath-to-mpi-headers -c test.cpp
g++ test.o -Lpath-to-mpi-libs -lmpi -lmpi_cxx -o test.out
```
Run with:
```
mpirun -np 2 ./test.out
```
I attached the valgrind output:
[message.txt](https://github.com/open-mpi/ompi/files/3281960/message.txt)

**Comment by jsquyres** [comment]
If you run with the latest 4.0.x snapshot tarball, does the same problem occur?

See https://www.open-mpi.org/nightly/v4.0.x/

**Comment by ggouaillardet** [comment]
This could be a duplicate of #6524 whose fix is available in the `v4.0.x` nightly tarball.

**Comment by thetwom** [comment]
Thanks, the nightly (I tested the Jun 08-version) indeed fixes the issue.
